### PR TITLE
Germany

### PR DESCRIPTION
**Pull request by joos2010kj**
- Add new params: run name, unique, and project name
    - Fix the germany and south africa's S1S2 dataset indexing issue
    - Run model with germany dataset (baseline)
    - Modify commands for Germany submission

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -143,7 +173,7 @@
 clip_value = 1e-2
 config["clip_value"] = clip_value
 for p in model.parameters():
-    if p.requires_grad:
+    if p.requires_grad and config['competition'] == 'south_africa':
```
Comment (joos2010kj) [inline train.py]: loss is nan if this condition is in place for germany

File: src/utils/s1_s2_reader.py [hunk 2]
```diff
@@ -71,9 +71,7 @@ def __init__(
             include_cloud=include_cloud,
         )
 
-        assert self.s1_reader.labels.drop("path", axis=1).equals(
-            self.s2_reader.labels.drop("path", axis=1)
-        )
+        # assert self.s1_reader.labels.equals(self.s2_reader.labels)
```
Comment (ivanzvonkov) [inline src/utils/s1_s2_reader.py]: Why the comment?

File: submit.py [hunk 3]
```diff
@@ -98,10 +104,11 @@
 
 output_frame = pd.DataFrame.from_dict(output_list)
 
-submission_path = Path(f"submissions/{name}/34S-20E-259N-2017-submission-{name}.json")
+submission_path = Path(f"submissions/{name}/{pos}-submission-{name}.json")
 submission_path.parent.mkdir(parents=True, exist_ok=True)
 output_frame.to_json(submission_path)
 
+import ipdb; ipdb.set_trace()
```
Comment (ivanzvonkov) [inline submit.py]: I think this line will fail inside sbatch

File: train.py [hunk 4]
```diff
@@ -55,11 +62,15 @@
...
 arg_parser.add_argument("--include_rvi", type=bool, default=False)
+arg_parser.add_argument("--alignment", type=str, default="1to2", help="Can be: 1to2 or 2to1 (76 vs. 41)")
+
+# WandB params
 arg_parser.add_argument("--disable_wandb", dest="enable_wandb", action="store_false")
 arg_parser.set_defaults(enable_wandb=True)
-arg_parser.add_argument(
-    "--alignment", type=str, default="1to2", help="Can be: 1to2 or 2to1 (76 vs. 41)"
-)
+arg_parser.add_argument("--name", type=str, default=None, help="Manually the run name (e.g., snowy-owl-10); None for automatic naming.")
```
Comment (ivanzvonkov) [inline train.py:76]: What's the use case for naming runs?

File: train.py [hunk 5]
```diff
@@ -26,9 +28,14 @@
...
-competition = "ref_fusion_competition_south_africa"
+
 arg_parser = ArgumentParser()
-arg_parser.add_argument("--competition", type=str, default=competition)
+arg_parser.add_argument(
+    "--competition", 
+    type=str, 
+    default="germany",
+    help="germany, south_africa"
+)
```
Comment (ivanzvonkov) [inline train.py:38]: If the original wandb project is for south africa then it would make sense for the default competition to be south africa too probably

File: train.py (hunk 1 above)
Comment (ivanzvonkov) [inline train.py]: What's going on here?

File: train.py [hunk 6]
```diff
@@ -162,7 +192,8 @@
 if config["enable_wandb"]:
     run = wandb.init(
         entity="nasa-harvest",
-        project="ai4food-challenge",
+        project=config['project'],
+        name=config['name'],
```
Comment (ivanzvonkov) [inline train.py:222]: Is this optional? Ie will it work if not specified?

File: train.py [hunk 7]
```diff
@@ -91,11 +118,13 @@
...
 )
 
-if config["pos"] == "both":
+if config["pos"] == "both_34":
     label_names_258, reader_258 = load_reader(pos="34S_19E_258N", **kwargs)
     print("\u2713 Loaded 258")
     label_names_259, reader_259 = load_reader(pos="34S_19E_259N", **kwargs)
     print("\u2713 Loaded 259")
+    reader_258.labels.reset_index(inplace=True)
+    reader_259.labels.reset_index(inplace=True)
```
Comment (ivanzvonkov) [inline train.py:138]: 1. Why not reset the index after concatenation?
2. Why does it work for South Africa data as is, but this change is needed for the Germany data?

File: src/utils/s1_s2_reader.py (hunk 2 above)
Comment (joos2010kj) [reply src/utils/s1_s2_reader.py]: Forgot.  Removed.

File: submit.py (hunk 3 above)
Comment (joos2010kj) [reply submit.py]: Right, forgot.  Removed.

File: train.py (hunk 4 above)
Comment (joos2010kj) [reply train.py:76]: It's been pretty helpful for me in identifying the run in wandb sweep.  Yes, you can rename it on the website, but I found naming it before sending it off more convenient.

File: train.py (hunk 7 above)
Comment (joos2010kj) [reply train.py:138]: 1. As mentioned in slack, I have to place the reset_index at those three places for fixing the iloc / loc issue.  And no, just putting one reset_index at the end of the whole if-else block doesn’t work because of the nature of ConcatDataset — it references reader_258 and reader_259 directly (those are not touched other than for accessing elements) and ConcatDataset keeps a different index series than those two.

More specifically, when you print ConcatDataset, you will see that its whole dataset is indexed sequentially (i.e., len = index), even if their subset datasets were initially not indexed sequentially (i.e., len < index).  Thus, resetting the index for the concatenated dataset has no impact.  However, if you apply reset_index to those two subsets BEFORE concatenation, then it's good.  Unless the index is the same as FID, reset_index will not have an issue.

File: train.py (hunk 6 above)
Comment (joos2010kj) [reply train.py:222]: Yes, the new params I added (--name, --unique, --project) all work as how it did before if left unspecified.  In other words, it will assign a random name and send it off to your wandb storage, not mine.

File: train.py (hunk 7 above)
Comment (joos2010kj) [reply train.py:138]: 2. < Give me a second >

**Review by joos2010kj — COMMENTED** [review]
Updated in the following commit and answered comments

File: train.py (hunk 4 above)
Comment (ivanzvonkov) [reply train.py:76]: what to you mean identifying? What do you name them?

File: train.py (hunk 1 above)
Comment (ivanzvonkov) [reply train.py]: > this clipping is needed for Germany dataset

Right now this code disables clipping for the Germany dataset, unless I am misunderstanding something

> loss is nan

Do you know why this is?

File: train.py (hunk 4 above)
Comment (ivanzvonkov) [reply train.py:76]: Ok it keeps the same behavior I am okay with this

File: train.py (hunk 7 above)
Comment (ivanzvonkov) [reply train.py:138]: This makes sense

File: train.py (hunk 1 above)
Comment (ivanzvonkov) [reply train.py]: I think this is the only thing that needs to be resolved!

Comment (ivanzvonkov) [reply train.py]: Any update on this?

Comment (joos2010kj) [reply train.py]: I've been working on getting the LSTM implementation first, so I'll get back to it once it's over.

**Comment by ivanzvonkov** [comment]
Okay, looks good. Please try the npz delete we talked about, that may eliminate the need for the filter!

**Comment by ivanzvonkov** [comment]
Is this good to merge?

**Comment by joos2010kj** [comment]
~Let me take a look at the merge conflicts now and you review the changes~

File: helper.py [hunk 8]
```diff
@@ -55,6 +55,13 @@ def load_reader(
 
     label_file = f"{root}/{country}_{train_or_test}_labels/{country}_{train_or_test}_labels_{pos}/labels.geojson"
     labels = gpd.read_file(label_file)
+
```
Comment (joos2010kj) [inline helper.py:58]: Needed for nan/inf value filtering from S1, S1_S2, and S1_S2_PlanetFusion

File: src/s1g_redflag.txt [hunk 9]
```diff
@@ -0,0 +1,85 @@
+52644
```
Comment (joos2010kj) [inline src/s1g_redflag.txt:1]: List of fids in germany sentinel 1 dataset containing inf/nan.  No access to test set, so can't remove them from test set.

File: train.py [hunk 10]
```diff
@@ -181,7 +183,7 @@
 clip_value = 1e-2
 config["clip_value"] = clip_value
 for p in model.parameters():
-    if p.requires_grad and config['competition'] == 'south_africa':
+    if p.requires_grad:
```
Comment (joos2010kj) [inline train.py]: Reverted back to original

**Review by joos2010kj — COMMENTED** [review]
Just a few comments in case difficult to follow.  ~LMK if it looks good one last time and I'll merge.~

**Comment by joos2010kj** [comment]
I will have to review codes again since it's crashing after merging when I run the code.

**Comment by ivanzvonkov** [comment]
Let me know if you need a second look

**Review by ivanzvonkov — APPROVED** [review]
